### PR TITLE
Creates librpc. Depends on #3504

### PR DESCRIPTION
**Pull request by xf6wang**
Separate RPC logic into a standalone library. 
`/mgmt/rpc/librpc_c.la (rpc client)`
`/mgmt/rpc/librpc_s.la (rpc server)`

Depends on #3504. Separated so it can be reviewed easier. 

Main Notes:
- Callbacks in the form `TSMgmtError(int, void*, size_t) ` can be registered. 
- Callbacks are now executed using via `Continuations` and `Events`. 

General Usage: 
- Clients should utilize `client_request` and `client_get_response` interface to interact with the RPC server. Every RPC request should contain a fd for the server, a int to indicate the function call followed by the parameters needed to execute the rpc call. 
- RPCCallbacks should expect the parameters to arrive in a void* buffer. All callbacks should use `mgmt_marshall_parse` to read off the parameters. 

Notes:
All client requests and responses begin with a integer to indicate the function. This is stripped away by the library and used to find the correct callback function.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
Is it really necessary to have a client vs. server library? One of the goals here was to remove the asymmetry of the RPC.

File: cmd/traffic_manager/traffic_manager.cc [hunk 1]
```diff
@@ -69,6 +70,7 @@
 // These globals are still referenced directly by management API.
 LocalManager *lmgmt = nullptr;
 FileManager *configFiles;
+RPCServerController *mgmt_server;
```
Comment (SolidWallOfCode) [inline cmd/traffic_manager/traffic_manager.cc]: Make this a `std::unique_ptr`.

File: cmd/traffic_manager/traffic_manager.cc [hunk 2]
```diff
@@ -680,30 +682,54 @@ main(int argc, const char **argv)
...
+
+  // Setup RPC mechanism.
+  mgmt_server = new RPCServerController(newmode);
+
+  if (mgmt_server->bindSocket(MGMTAPI_MGMT_SOCKET_NAME) == -1) {
+    // mgmt_log("[WebIntrMain] Unable to set up socket for handling management API calls. API socket path = %s\n", apisock.c_str());
+    lmgmt->alarm_keeper->signalAlarm(MGMT_ALARM_WEB_ERROR, mgmtapiFailMsg);
+  }
+
+  mgmt_server->registerControlCallback((RECORD_SET), handle_record_set);
```
Comment (SolidWallOfCode) [inline cmd/traffic_manager/traffic_manager.cc]: Why the extra set of parens?

File: mgmt/BaseManager.h [hunk 3]
```diff
@@ -83,27 +83,27 @@
...
-#define MGMT_SIGNAL_MACHINE_DOWN 2
-#define MGMT_SIGNAL_CONFIG_ERROR 3 /* Data is descriptive string */
-#define MGMT_SIGNAL_SYSTEM_ERROR 4
-#define MGMT_SIGNAL_LOG_SPACE_CRISIS 5
-#define MGMT_SIGNAL_CONFIG_FILE_READ 6
-#define MGMT_SIGNAL_CACHE_ERROR 7
-#define MGMT_SIGNAL_CACHE_WARNING 8
-#define MGMT_SIGNAL_LOGGING_ERROR 9
-#define MGMT_SIGNAL_LOGGING_WARNING 10
+#define MGMT_SIGNAL_PID 100
```
Comment (SolidWallOfCode) [inline mgmt/BaseManager.h]: These need to be unified, it's a bad idea to have effectively the same values in different headers.

File: proxy/Makefile.am [hunk 4]
```diff
@@ -40,6 +40,7 @@ AM_CPPFLAGS += \
   -I$(abs_srcdir)/hdrs \
   -I$(abs_srcdir)/shared \
   -I$(abs_top_srcdir)/mgmt \
+  -I$(abs_top_srcdir)/mgmt/rpc \
```
Comment (SolidWallOfCode) [inline proxy/Makefile.am]: Don't do this - use `#include <rpc/...>` instead. We're trying to reduce the massive list of include directories.

File: mgmt/LocalManager.cc [hunk 5]
```diff
@@ -578,19 +578,26 @@ LocalManager::handleMgmtMsgFromProcesses(MgmtMessageHdr *mh)
     break;
   // Congestion Control - end
   case MGMT_SIGNAL_CONFIG_FILE_CHILD: {
-    static const MgmtMarshallType fields[] = {MGMT_MARSHALL_STRING, MGMT_MARSHALL_STRING, MGMT_MARSHALL_INT};
-    char *parent                           = nullptr;
-    char *child                            = nullptr;
-    MgmtMarshallInt options                = 0;
-    if (mgmt_message_parse(data_raw, mh->data_len, fields, countof(fields), &parent, &child, &options) != -1) {
+<<<<<<< HEAD
```
Comment (zwoop) [inline mgmt/LocalManager.cc]: This looks wrong (merge conflicts)?

**Comment by xf6wang** [comment]
@zwoop I'm waiting for #3504 to go in before squashing, I think the others commit will disappear. I just separated the commits for reviewing. This PR is basically just for "Move RPC to separate lib". I rebased again, are there still merge conflicts?

File: proxy/Makefile.am (hunk 4 above)
Comment (xf6wang) [reply proxy/Makefile.am]: updated.

File: mgmt/BaseManager.h (hunk 3 above)
Comment (xf6wang) [reply mgmt/BaseManager.h]: I think there supposed to be the exact same values. There's a comment "copy from mgmt/BaseManager.h"

File: cmd/traffic_manager/traffic_manager.cc (hunk 2 above)
Comment (xf6wang) [reply cmd/traffic_manager/traffic_manager.cc]: Consequence of doing a search and replace when I was removed `OpType`. Removed.

File: cmd/traffic_manager/traffic_manager.cc (hunk 1 above)
Comment (xf6wang) [reply cmd/traffic_manager/traffic_manager.cc]: done.

**Comment by xf6wang** [comment]
@SolidWallOfCode I have two separate libraries because using `Continuations` pulls in `/iocore/eventsystem/libinkevent.a` which consequently needs `/lib/records/librecords_lm.a` which then needs `/mgmt/libmgmt_c.la`. So, without the split, all the other executables, (ie `traffic_cop`) need a bunch of extra libraries.

**Comment by zwoop** [comment]
[approve ci clang-format]

**Comment by SolidWallOfCode** [comment]
It was determined at the Spring 2018 summit to take the different approach of using a third party serialization library. We will still create a `libRPC` based on that.

**Pull request closed without merging**